src/index.ts

```diff
@@ -45,21 +48,27 @@ const createGroups = function (Bottleneck, common) {
   });
 };
 
-export function throttling(octokit: Octokit, octokitOptions = {}) {
+export function throttling(
+  octokit: Octokit,
+  options: { throttle: ThrottlingOptionsBase & SecondaryLimitHandler }
+): {};
+export function throttling(
+  octokit: Octokit,
+  options: { throttle: ThrottlingOptionsBase & AbuseLimitHandler }
```

Comment:
Those are [TS method overloads](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads) to make the typing work, do you suggest to remove them?

I tried to do it without that and I was not getting the expected typing: 
<img width="257" alt="image" src="https://user-images.githubusercontent.com/2574275/157774540-668e9f63-b1b4-4b8e-9f04-6f8b0068c75c.png">


See [this TS playground ](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgeQMYwga2PAvnAMyghDgCIABCdLHAelWgFMyBuAKFEljgCEIYMADZMAdk1SZCxUmQBGA4WImY27TqJhMoBAIaomKGthjIwMYBFEBnROzgO4AbQjnLogFxxrMKMFEA5gC6XrqiAJ4cjnC6AK4wABYAyr66WgHhAPyhEVGOcYnZMbn2jrHW2gCCAWIwRT5+gXAAPnCxogAmTAT+TB15DmBQTABuwEwA7tb1vv4BTkEtbZ3dvf2lDnK6FQCqUEIzjQFL7V094uvRQhABRQgb0XBdcrEBXgAUIEzW1ro1Xg05gBKOAAXgAfMtMKIIBNRANHv4CBAPl8fn8mADZoEQRCoTC4QjohNdFBPHBPt9fv9vNiArjIe1obD4Q9otpiFBUVSMVijgz8SyEfhWqdVhcEcMAI6xb51HKRB4WL4ALSsTEOcxOK3OfQ4uHU-i0On0hjQGBMZgsVls9wciWIghERQAKglHcI5lb3NZ9eoYOEwIYADKgHAACTCHREUDBFNKw184UqBGNRVEsRAcm0ABpSq5rTYihA5AArCQwPMOagWnBFc20GDsAUjCDAdbsANBuCVF4VUMgCNRmNxu1wOgAKgnGwncAoXSGEjSfXITioxnoYCEr38AFoHYohHNFgADKy98pMAc4d5Ak9wYC2BfDVDLjrFd8TYBCIRwbNwYYQAgEYV38GJCHiWJhgApgRG2QwIAIOAT3XWsYDoLcd1Efd3UPOYTxzOAtyYeC2gqOBEkMM9RCSCQrA6UlwgAJWXa8YHvBJh20B8bC0XQOjIGc6HzUQL37MMYC8NjI06GN9Q4TtA0MWjGE6RjpK42NQTsasaLotSoGY1iJKkiSZOjbR5PUJgAA9uHgLtDDdD0j0Cb0bV4UjtLHMRdDkEQOiKBQIDg1kHH4J1lEkIowkVBx201QI8mVJgIHidNM2zKA8lU8R0HcGKSgcQdRFATMVPoxiWK0JimCTFM0y8DMs0shM6sMhrtE8ioADVdG3DUmsy1qHAmPwtCKCKlDyzAADoAHFiFiMA8gqUlUASSbDyiubFrSlbShhCwelfQtpi8KaRBmhaloO3TqqvEy4A02TWtwBTHLgZy8LctwbTjb6nS9P6bG6wwADJSneMTHsHGBzJHVoKoM8IXosqAgQUgh2nyqw4E416oHeGtGwVQiCx9LwG0tEHrBBMdEyg1kDXYOg6AoGBrF3Wyg3QbmoE5dgCfR94EFwQixwPSKvDHe7jLhj4BTFqtHF8-y+i8XxZRV3SYbYxWwUhZWHisZGGMMh79YpJWDQcA1cCBdRhZjUXxbsNmOa5nmK35wX7Vw6WdMcKxLae29DcQcWHjVgLNagbXSntx2hc012JdKKWlBlk3RFDhXrYj432VEPzY4o+OmB1uBzz7WGcANvEi7gJOndT5Wg8zkRs+iEP5frgvG6j4vS418uE57vTVPNoytCt8PB5VpOgA)

Any better suggestion or something I'm missing?

Reply:
I didn't notice that they were overloads.

I removed them, and it works for me. Are you sure you patched your `node_modules`?

Reply:
> I didn't notice that they were overloads.
> 
> I removed them, and it works for me. Are you sure you patched your `node_modules`?

This is the file I modified:

<img width="532" alt="image" src="https://user-images.githubusercontent.com/2574275/157778439-9c4618d2-48f4-49f1-a88c-7e50ced9e260.png">

Is the right one, right?

Also, in the TS Playground I attached before, I'm "hardcoding" the double interface declaration but it does not work. Is not a good example to check?

Reply:
That is the file that should be modified, mine looks the same.

This is what the `throttling` function looks like for me:
![image](https://user-images.githubusercontent.com/4595477/157778992-cca0d076-e7b0-4be2-b1a6-fa479c093637.png)

When I hover the `octokitOptions.throttle` option, I get the proper types, as seen in the image
![image](https://user-images.githubusercontent.com/4595477/157778814-7061af8a-f8ab-4342-b2e7-57e6d0ee6562.png)

Reply:
This is the complete diff of my changes
```diff
diff --git a/src/index.ts b/src/index.ts
index 99e4d91..ce1b561 100644
--- a/src/index.ts
+++ b/src/index.ts
@@ -48,14 +48,6 @@ const createGroups = function (Bottleneck, common) {
   });
 };
 
-export function throttling(
-  octokit: Octokit,
-  options: { throttle: ThrottlingOptionsBase & SecondaryLimitHandler }
-): {};
-export function throttling(
-  octokit: Octokit,
-  options: { throttle: ThrottlingOptionsBase & AbuseLimitHandler }
-): {};
 export function throttling(octokit: Octokit, octokitOptions: OctokitOptions) {
   const {
     enabled = true,
@@ -113,7 +105,7 @@ export function throttling(octokit: Octokit, octokitOptions: OctokitOptions) {
   events.on(
     "secondary-limit",
     isUsingDeprecatedOnAbuseLimitHandler
-      ? function (...args: [number, ThrottlingOptions, Octokit]) {
+      ? function (...args: [number, OctokitOptions, Octokit]) {
           octokit.log.warn(
             "[@octokit/plugin-throttling] `onAbuseLimit()` is deprecated and will be removed in a future release of `@octokit/plugin-throttling`, please use the `onSecondaryRateLimit` handler instead"
           );
```

Reply:
I just added a new file `test/typescript-validate.ts` (inspired by the one we have in [@octokit/webhooks.js](https://github.com/octokit/webhooks.js/blob/master/test/typescript-validate.ts)) so we can validate every time we modify tests are not introducing a regression.

Reply:
I just pushed this branch: `improve-types-with-wolfy-suggestion` with your changes.
Also, when doing cmd+click to `OctokitOptions` there is what I get so I understand I'm updating `OctokitOptions` properly yn my `node_modules`:

![image](https://user-images.githubusercontent.com/2574275/158036366-72d55c66-6a68-4990-bc47-697c7fd4f4ec.png)

What I obtain as unexpected TS errors are these ones:
![image](https://user-images.githubusercontent.com/2574275/158036373-ef4ff903-68eb-432c-a9ca-baeb7e794105.png)

@wolfy1339 can you check if those type validations work in your machine?

Reply:
The type tests work as expected, the same 2 fail for me.

The `// @ts-expect-error` flags should be above the `throttle` key of the options object, in order to silence the error.

Reply:
> The type tests work as expected, the same 2 fail for me.
> 
> The `// @ts-expect-error` flags should be above the `throttle` key of the options object, in order to silence the error.

Yes, for the second case in the screenshot yes, you are right. For the first one, in my local, TS does not complain in `improve-types-with-wolfy-suggestion` branch as expected. Does it complain for you?

Reply:
The `onSecondaryLimit() and onAbuseLimit() should be a TS Error` test does not complain for me.

Reply:
Can I see how your modified `node_modules` file looks like?

This is mine:

<details>
  <summary>`node_modules/@octokit/core/dist-types/types.d.ts`</summary>


```javascript
export declare interface OctokitOptions {
    authStrategy?: any;
    auth?: any;
    userAgent?: string;
    previews?: string[];
     baseUrl?: string;
    log?: {
        debug: (message: string) => unknown;
        info: (message: string) => unknown;
        warn: (message: string) => unknown;
        error: (message: string) => unknown;
    };
    request?: OctokitTypes.RequestRequestOptions;
    timeZone?: string;
    [option: string]: any;
}
  ```
</details>

Reply:
```ts
export declare interface OctokitOptions {
    authStrategy?: any;
    auth?: any;
    userAgent?: string;
    previews?: string[];
    baseUrl?: string;
    log?: {
        debug: (message: string) => unknown;
        info: (message: string) => unknown;
        warn: (message: string) => unknown;
        error: (message: string) => unknown;
    };
    request?: OctokitTypes.RequestRequestOptions;
    timeZone?: string;
    [option: string]: any;
}```

Reply:
Same file (check path)

Reply:
`node_modules/@octokit/core/dist-types/types.d.ts`

Reply:
The version of TS that VSCode uses, and the one present in the repo, are 2 different versions.

VSCode: `4.6.2`
Repo: `4.4.3`

Reply:
> `node_modules/@octokit/core/dist-types/types.d.ts`

Strange. I guess we will need to sole this issue (https://github.com/octokit/core.js/issues/451), update this branch with the new version, apply your suggestion and push it here so CI tell us if it works.

Reply:
@wolfy1339 updated this branch with `@octokit/core` and CI is running `validate:ts` now. Seems it is getting the same error I'm having. When you run `npm run validate:ts` in your local, does it work?

In case it doesn't, I guess we will need to rethink the typing (maybe with the approach I was trying to use?)

Reply:
The `onSecondaryLimit() and onAbuseLimit() should be a TS Error` still doesn't error for me, with clean `node_modules`.
```
$ npm run validate:ts

> @octokit/plugin-throttling@0.0.0-development validate:ts
> tsc --noEmit --noImplicitAny --target es2020 --esModuleInterop --moduleResolution node test/typescript-validate.ts

test/typescript-validate.ts:26:3 - error TS2578: Unused '@ts-expect-error' directive.

26   // @ts-expect-error
     ~~~~~~~~~~~~~~~~~~~


Found 1 error.
```

Reply:
@G-Rath Would you have time to assist us on this?

Right now we are having problems making the types a bit more rigid.

The only way I know to make them more rigid is using method overload, but this makes [TS complain](https://github.com/octokit/plugin-throttling.js/pull/457#discussion_r824273665). Is there a different way to make the types more rigid and at the same time using the Declaration merging you suggested?

Reply:
Try this:

```
export type AbuseLimitHandler = {
  /**
   * @deprecated "[@octokit/plugin-throttling] `onAbuseLimit()` is deprecated and will be removed in a future release of `@octokit/plugin-throttling`, please use the `onSecondaryRateLimit` handler instead"
   */
  onAbuseLimit: LimitHandler;
  onSecondaryRateLimit?: never;
};

export type SecondaryLimitHandler = {
  onAbuseLimit?: never;
  onSecondaryRateLimit: LimitHandler;
};
```

iirc this is because objects in TypeScript are "open", so `x | y` means that the object must satisfy at least the type for `x | y`, not that it can't satisfy the type of both.

You can (sort of*) enforce this by using optional + `never`: since each type now defines the inverse property as having to have type `never`, defining both results in an error because such an object does not satisfy either type in the union.

*: from what I understand this is technically not the same as what we really want, e.g. it's still defining an actual property, so IDEs might still include it in their autocomplete lists, and TypeScripts errors about this will be something like "this property should be type never" rather than a more helpful "you can't implement this property _and_ that property", but it should be close enough to what you want (and it's as close as you're able to get currently anyway)

Reply:
Thanks once again @G-Rath. Since this will be a temporal typing until we deprecate `onAbuse` I think is more than enough.